src/runtime_src/core/pcie/driver/linux/xocl/subdev/icap.c

```diff
@@ -2202,6 +2203,23 @@ static int __icap_xclbin_download(struct icap *icap, struct axlf *xclbin, bool s
 		err = icap_refresh_clock_freq(icap, xclbin);
 		if (err)
 			ICAP_ERR(icap, "not able to refresh clock freq");
+	} else if (err == -EEXIST) {
+		/* Try locating the ep_freq_cnt_aclk_kernel_* endpoints in xclbin
+		 * On u2 raptor2 shells (1RP or 2RP or 0RP), these endpoints are
+		 * available in ULP (i.e. in xclbin) but not in BLP+PLP
+		 */
+		const struct axlf_section_header *header = NULL;
+		header = xrt_xclbin_get_section_hdr(xclbin, PARTITION_METADATA);
+		if (header) {
+			struct clock_counter_info *clk_counter;
+			clk_counter = (struct clock_counter_info *) vzalloc(CCT_NUM *
```

Comment:
alternatively, we could avoid using vzalloc and vfree by 
struct clock_counter_info clk_counter = { 0 };
then pass &clk_counter into the arg